packages/ckeditor5-dev-license-checker/src/validate-license-files.ts

```diff
@@ -0,0 +1,351 @@
+	fix = false,
+	shouldProcessRoot = false,
+	shouldProcessPackages = false,
+	isPublic = false,
+	rootDir,
+	projectName,
+	mainPackageName,
+	copyrightOverrides = []
+}: {
+	fix?: boolean;
+	shouldProcessRoot?: boolean;
+	shouldProcessPackages?: boolean;
+	isPublic?: boolean;
+	rootDir: string;
+	projectName: string;
+	mainPackageName?: string;
+	copyrightOverrides?: Array<CopyrightOverride>;
+} ): Promise<number> {
+	const packagePaths: Array<string> = [];
+
+	if ( shouldProcessRoot ) {
+		packagePaths.push( rootDir );
+	}
+
+	if ( shouldProcessPackages ) {
+		packagePaths.push( ...await fromAsync( glob( upath.join( rootDir, 'packages', '*' ) ) ) );
+	}
+
+	if ( !packagePaths.length ) {
+		console.error( [
+			'No packages to parse detected. Make sure that you provided proper paths,',
+			'as well as set at least one of: `shouldProcessRoot` or `shouldProcessPackages`.'
+		].join( '\n' ) );
+
+		return 1;
+	}
+
+	// Collect versioning and licensing data of all dependencies.
+	const dependencyMap = await Promise.all( packagePaths.map( async packagePath => {
+		const pkgJsonPath = upath.join( packagePath, 'package.json' );
+		const pkgJsonContent = JSON.parse( await readFile( pkgJsonPath, 'utf-8' ) );
+		const dependencyNames = Object.keys( pkgJsonContent.dependencies || {} )
+			.filter( dependency => !dependency.match( /(ckeditor)|(cksource)/i ) );
+
+		const packageName = pkgJsonContent.name;
+
+		const dependencyMapItem: DependencyMapItem = {
+			packageName,
+			packagePath,
+			dependencies: []
+		};
+
+		const copyrightOverridesPackage = copyrightOverrides
+			.find( ( { packageName: overridePackageName } ) => packageName === overridePackageName );
+
+		if ( copyrightOverridesPackage ) {
+			dependencyMapItem.dependencies.push( ...copyrightOverridesPackage.dependencies );
+		}
+
+		for ( const dependencyName of dependencyNames ) {
+			// If override already exists, skip parsing the dependency.
+			if ( dependencyMapItem.dependencies.some( ( { name } ) => name === dependencyName ) ) {
+				continue;
+			}
+
+			const dependencyData: DependencyMapItem['dependencies'][number] = { name: dependencyName };
+			dependencyMapItem.dependencies.push( dependencyData );
+
+			let dependencyPath: string;
+
+			try {
+				const fakeEntry = upath.join( packagePath, 'fakeEntry.js' );
+				const require = createRequire( fakeEntry );
+				const pkgJsonPath = require.resolve( `${ dependencyName }/package.json` );
+
+				dependencyPath = upath.dirname( pkgJsonPath );
+			} catch ( err: any ) {
+				// If `dependencyName/package.json` is not a valid import and throws an error, the error includes the path to the
+				// `package.json` that we need, and we can read it. This error catching mechanism is also needed to find paths to packages
+				// which do not have a base export, eg. package `empathic` only has exports such as `empathic/find`.
+				const dependencyPkgJsonPath = err?.message.match( /(?<=not defined by "exports" in ).+(?=\s|$)/ )?.[ 0 ];
```

Comment:
Depending on specific text in error messages is a bad idea, because they tend to change without notice. Recently something similar happened in Firefox, which broke similar implementations.

I think going with [`findPackageJSON`](https://nodejs.org/api/module.html#modulefindpackagejsonspecifier-base) is a much safer approach.